Deserializer for unexpected array format from endpoints 2

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-1037

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This adds support for parsing `"meta_data": {"1": {...}, "2": {...}}` when the expected format is `"meta_data": […]`. This unexpected format can sometimes be sent by the server due to PHP behavior. 
Previously, we added this for the `List<*>` type, but we're still seeing crashes, so we're now adding it for `Array<*>` type as well.  This was discussed in https://github.com/woocommerce/woocommerce-android/pull/14553#discussion_r2318959270

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
I don’t know how to reproduce this case. Reviewing the unit tests should be enough.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Introduce `ArrayOrObjectDeserializer`

This commit introduces a new Gson deserializer, `ArrayOrObjectDeserializer`.

irfano committed:
Add tests for `ArrayOrObjectDeserializer`

irfano committed:
Add `ArrayOrObjectDeserializer` to `GsonRequest`

hichamboushaba commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ArrayOrObjectDeserializer.kt:
```diff
@@ -0,0 +1,62 @@
+        typeOfT: Type,
+        context: JsonDeserializationContext
+    ): Array<*>? {
+        val componentType: Type = when (typeOfT) {
+            is GenericArrayType -> typeOfT.genericComponentType
+            is Class<*> -> if (typeOfT.isArray) {
+                typeOfT.componentType
+            } else {
+                throw JsonParseException("Expected array: $typeOfT")
+            }
+
+            else -> throw JsonParseException("Unsupported type: $typeOfT")
+        }
```
np, as this will be called only for arrays, we can be sure that the type here is `GenericArrayType`, so we can simplify this to something like this:

```kotlin
        val componentType: Type = (typeOfT as? GenericArrayType)?.genericComponentType
            ?: throw JsonParseException("Expected a GenericArrayType, but got: $typeOfT")
```

irfano replied:
Good catch! Your patch makes it simpler and better.

hichamboushaba commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ArrayOrObjectDeserializer.kt:
```diff
@@ -0,0 +1,62 @@
+
+            else -> throw JsonParseException("Unsupported type: $typeOfT")
+        }
+        val raw: Class<*> = when (componentType) {
+            is Class<*> -> componentType
+            is ParameterizedType -> componentType.rawType as Class<*>
+            else -> Any::class.java
+        }
```
np, we can use the `GsonTypes.getRawType` function to get the raw type directly.

irfano replied:
Good advice!

hichamboushaba reviewed (approved):
Nice work @irfano, tests work as expected, and ready to ship.

I'm just leaving an optional suggestion for you, I think we can simplify the code a bit and reduce the code repetition between this new component and the existing `ListOrObjectDeserializer`, please check the following patch:

```patch
Index: libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ArrayOrObjectDeserializer.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ArrayOrObjectDeserializer.kt b/libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ArrayOrObjectDeserializer.kt
--- a/libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ArrayOrObjectDeserializer.kt	(revision 88217ddda7dd0897c76a173d49272a1e6133797e)
+++ b/libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ArrayOrObjectDeserializer.kt	(date 1758795060680)
@@ -4,9 +4,11 @@
 import com.google.gson.JsonDeserializer
 import com.google.gson.JsonElement
 import com.google.gson.JsonParseException
+import com.google.gson.internal.GsonTypes.getRawType
 import java.lang.reflect.GenericArrayType
 import java.lang.reflect.ParameterizedType
 import java.lang.reflect.Type
+import java.lang.reflect.Array as JArray
 
 /**
  * Deserializes a JSON value that can be either an array or an object with numeric string keys.
@@ -27,36 +29,68 @@
         typeOfT: Type,
         context: JsonDeserializationContext
     ): Array<*>? {
-        val componentType: Type = when (typeOfT) {
-            is GenericArrayType -> typeOfT.genericComponentType
-            is Class<*> -> if (typeOfT.isArray) {
-                typeOfT.componentType
-            } else {
-                throw JsonParseException("Expected array: $typeOfT")
-            }
+        val componentType: Type = (typeOfT as? GenericArrayType)?.genericComponentType
+            ?: throw JsonParseException("Expected a GenericArrayType, but got: $typeOfT")
+
+        val items: List<Any?> = jsonToList(json, componentType, context)
+
+        return items.listToArray(componentType)
+    }
+
+    private fun List<Any?>.listToArray(componentType: Type): Array<Any?> {
+        val arr = JArray.newInstance(getRawType(componentType), size)
+        forEachIndexed { i, v -> JArray.set(arr, i, v) }
+        @Suppress("UNCHECKED_CAST")
+        return arr as Array<Any?>
+    }
+}
 
-            else -> throw JsonParseException("Unsupported type: $typeOfT")
-        }
-        val raw: Class<*> = when (componentType) {
-            is Class<*> -> componentType
-            is ParameterizedType -> componentType.rawType as Class<*>
-            else -> Any::class.java
+/**
+ * Deserializes a JSON value that can be either an array or an object with numeric string keys.
+ * Examples:
+ * - Array: ["item1", "item2"]
+ * - Object with numeric keys: {"0": "item1", "1": "item2"}
+ * - Non-ordered numeric keys example: {"2":"a","1":"b","10":"z"} -> ["b", "a", "z"]
+ *
+ * This deserializer handles both cases and converts them into a List<T>.
+ * - For arrays: elements are deserialized in their natural order.
+ * - For objects: keys must be numeric strings (0, 1, 2, ...). Values are read by sorting keys numerically
+ * (0, 1, 2, ...), not by insertion order. Any non-numeric key results in a JsonParseException.
+ * If the JSON value is null, null is returned.
+ */
+class ListOrObjectDeserializer : JsonDeserializer<List<*>> {
+    override fun deserialize(
+        json: JsonElement,
+        typeOfT: Type,
+        context: JsonDeserializationContext
+    ): List<*>? {
+        val itemType = (typeOfT as ParameterizedType).actualTypeArguments[0]
+        return jsonToList(json, itemType, context)
+    }
+}
+
+private fun jsonToList(
+    json: JsonElement,
+    itemType: Type,
+    context: JsonDeserializationContext
+): List<Any?> {
+    return when {
+        json.isJsonArray -> json.asJsonArray.map { element ->
+            context.deserialize(element, itemType)
         }
-        val items: List<Any?> = when {
-            json.isJsonArray -> json.asJsonArray.map { context.deserialize(it, componentType) }
-            json.isJsonObject -> json.asJsonObject.entrySet()
-                .onEach {
-                    it.key.toIntOrNull() ?: throw JsonParseException("Unexpected key: ${it.key}")
-                }
-                .sortedBy { it.key.toInt() }
-                .map { context.deserialize(it.value, componentType) }
-
-            json.isJsonNull -> return null
-            else -> throw JsonParseException("Unexpected JSON for Array: $json")
-        }
-        val arr = java.lang.reflect.Array.newInstance(raw, items.size)
-        items.forEachIndexed { i, v -> java.lang.reflect.Array.set(arr, i, v) }
-        @Suppress("UNCHECKED_CAST")
-        return arr as Array<*>
+
+        json.isJsonObject -> json.asJsonObject.entrySet()
+            .onEach { entry ->
+                val key = entry.key
+                key.toIntOrNull()
+                    ?: throw JsonParseException("Unexpected key in JSON object matching array: $key")
+            }
+            .sortedBy { it.key.toInt() }
+            .map { entry ->
+                context.deserialize(entry.value, itemType)
+            }
+
+        json.isJsonNull -> emptyList()
+        else -> throw JsonParseException("Unexpected JSON type for List, $json")
     }
 }
Index: libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ListOrObjectDeserializer.kt
===================================================================
diff --git a/libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ListOrObjectDeserializer.kt b/libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ListOrObjectDeserializer.kt
deleted file mode 100644
--- a/libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/ListOrObjectDeserializer.kt	(revision 88217ddda7dd0897c76a173d49272a1e6133797e)
+++ /dev/null	(revision 88217ddda7dd0897c76a173d49272a1e6133797e)
@@ -1,50 +0,0 @@
-package org.wordpress.android.fluxc.network.rest
-
-import com.google.gson.JsonDeserializationContext
-import com.google.gson.JsonDeserializer
-import com.google.gson.JsonElement
-import com.google.gson.JsonParseException
-import java.lang.reflect.ParameterizedType
-import java.lang.reflect.Type
-
-/**
- * Deserializes a JSON value that can be either an array or an object with numeric string keys.
- * Examples:
- * - Array: ["item1", "item2"]
- * - Object with numeric keys: {"0": "item1", "1": "item2"}
- * - Non-ordered numeric keys example: {"2":"a","1":"b","10":"z"} -> ["b", "a", "z"]
- *
- * This deserializer handles both cases and converts them into a List<T>.
- * - For arrays: elements are deserialized in their natural order.
- * - For objects: keys must be numeric strings (0, 1, 2, ...). Values are read by sorting keys numerically
- * (0, 1, 2, ...), not by insertion order. Any non-numeric key results in a JsonParseException.
- * If the JSON value is null, null is returned.
- */
-class ListOrObjectDeserializer : JsonDeserializer<List<*>> {
-    override fun deserialize(
-        json: JsonElement,
-        typeOfT: Type,
-        context: JsonDeserializationContext
-    ): List<*>? {
-        val itemType = (typeOfT as ParameterizedType).actualTypeArguments[0]
-        return when {
-            json.isJsonArray -> json.asJsonArray.map { element ->
-                context.deserialize<Any?>(element, itemType)
-            }
-
-            json.isJsonObject -> json.asJsonObject.entrySet()
-                .onEach { entry ->
-                    val key = entry.key
-                    key.toIntOrNull()
-                        ?: throw JsonParseException("Unexpected key in JSON object matching array: $key")
-                }
-                .sortedBy { it.key.toInt() }
-                .map { entry ->
-                    context.deserialize<Any?>(entry.value, itemType)
-                }
-
-            json.isJsonNull -> null
-            else -> throw JsonParseException("Unexpected JSON type for List, $json")
-        }
-    }
-}
```

irfano committed:
Refactor Deserializers

This commit refactors the `ArrayOrObjectDeserializer` and `ListOrObjectDeserializer` classes.

Additionally, `ListOrObjectDeserializer.kt` has been removed, and its contents were moved into `ArrayOrObjectDeserializer.kt` since they are closely related.
